design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,312 @@
+we should be in a pretty good shape, for most commands we return specific structs (defined in `cast/src/helpers/response_structs.rs`),
+but let's double-check everything necessary is included there, and all importable commands are using such structs.
+
+### sncast config
+We should allow for all flags passed to sncast to be propagated to the script. CastConfig should also be importable
+and usable from script if someone wishes so, but we should not require it. That means, we might have to review our subcommands
+and if some of them requires it directly (`account create` is one), we should try to get rid of this dependency. 
+
+### interacting with contract
+We will use contracts dispatchers to be able to call/invoke its functions directly 
+(e.g. if contract named `mycontract` has a function `myfunction`, we should be able to just do `mycontract.myfunction();`).
+Later on we could also support our subcommands (`invoke`, `call`) to call/invoke contracts without dispatchers.
+
+### running the script
+The script would essentially run an entrypoint function - `main`. Inside our script subcommand, we will have to compile 
+the cairo script file using `scarb build`, and then run it using `scarb cairo-run`. 
+The main function would be required in the deployment script, and we should return an error if it is not found.
+
+In order to build the script, there'll need to be a `Scarb.toml` file present in the working directory. We should either:
+- look for a Scarb.toml in the same directory script is (this is scarb's default behaviour - we just must make sure scarb
+command is executed from this directory - Command::new(...).current_dir(...) should do); if does not exist we could just create it
+- create a hidden Scarb.toml with a custom name (eg `.script_name.toml`) and override path to scarb toml using scarb flag;
+if a file does not exist let's just create it
+
+Option 1 is probably more in line with scarb ethos, but the option 2 would probably be more practical for us as it would allow
+for multiple Scarb.toml files in one directory, where multiple scripts could reside + Scarb.toml file present in cwd would not
+confuse sncast subcommands (eg declare also looks for Scarb.toml in cwd, which could lead to errors/user confusion).
+
+The deployment script could then be run like so:
+```bash
+$ sncast script /path/to/myscript.s.cairo
+```
+
+Log info should be outputted to stdout and optionally written to some log file. 
+We could have a flag to suppress outputting to stdout or to output with different formats (int, hex).
+
+By default, the scripts could run in some kind of "dry run mode", meaning all the transactions would be simulated with 
+specific rpc calls (like `starknet_simulateTransactions`). An interesting idea would be to research an approach similar 
+to cheatnet - fork a runner and test the scripts against it (to be assessed if it is feasible). We could also provide 
+a way to show a trace of said transactions if required.
```

Comment:
It would be nice to have such a flow:
- run a script
- create devnet instance
- fork the state
- run script against forked devnet

Reply:
I suppose this would look like this:

- someone runs devnet locally with forked testnet state
- in the deployment script uses `get_provider` with devnet url
- profit


would that be ok, or did you have something else in mind?

Reply:
I think it is fine and we should allow for both. My idea was to do it `under the hood` so a user would not have to setup devnet by itself

Reply:
this could be good, but I think we should wait for this for devnet-rs. Will add info about this to the doc.